topi/python/topi/mali/conv2d.py

```diff
@@ -188,21 +188,16 @@ def _schedule_spatial_pack(cfg, s, output, conv, data_vec, kernel_vec):
     return s
 
 ##### WINOGRAD TEMPLATE #####
-def _pick_tile_size(data, kernel):
-    N, CI, H, W = get_const_tuple(data.shape)
-
-    if H % 4 == 0:
-        return 4
-    else:
-        return 2
-
 @autotvm.register_topi_compute(conv2d, 'mali', ['winograd'])
 def conv2d_mali_winograd(cfg, data, kernel, strides, padding, dilation, layout, out_dtype):
-    tile_size = _pick_tile_size(data, kernel)
-    return _decl_winograd(cfg, data, kernel, strides, padding, dilation, layout, out_dtype,
-                          tile_size)
+    return _decl_winograd(cfg, data, kernel, strides, padding, dilation, layout, out_dtype)
+
+def _decl_winograd(cfg, data, kernel, strides, padding, dilation, layout, out_dtype,
+                   tile_size=None):
+
+    cfg.define_knob('tile_size', enum_tile_sizes(data))
```

Comment:
Maybe we should upgrade our `tune_relay_arm` / `tune_relay_mali` tutorial, we should make xgb tuner feature_type be `knob`. Otherwise, we maybe hang forever.

Reply:
@FrozenGene ,

 Are you sure about ```xgb``` vs. ```xgb_knob``` ? Cfg space is increasing only few times, its under one milion in size. In my experience ```xgb``` is still performing way better, in fact ```xgb_knob``` did not found anything useful. But I may wrong, please confirm / decline .

Reply:
`XGBTunner`'s default `feature_type` is `itervar`, see: https://github.com/dmlc/tvm/blob/master/python/tvm/autotvm/tuner/xgboost_tuner.py#L68. When we have `define_knob`, if we don't change `feature_type`from `itervar` into `knob`, we maybe will meet hang forever (i.e. can not tune and just hang). Not related with config search space. Just our feature exaction issue. I met this problem when I `define_knob` in depthwise convolution optimization on ARM CPU. cc @merrymercy

Reply:
@FrozenGene ,

 * Experienced those hangouts even crashes when cfg.space was huge like in #2111 (working on it). My feeling was that features get corrupted bad way and xgb gets confused (too large !?). Willing to fix that issue anyway i am working on bring #2111 in TVM finally in (hope to look/fix at ```xgb``` issue too). My test (so far) shows that too much amount of (options like knob) may cause hangs.
  * On other side, we already have knobs there (working fine) we just added another one.

 I triple check: still sure want to update related tutorials to ```xgb_knob``` ? I am confident it works just fine with plain ```xgb``` (and better), but if you still want, I firmly update tutorials (immediate, no more objections).

Reply:
Ok. Let's us keep `xgb`. Wish you bring #2111 in could have time to look at `xgb` issue too. :-)

Reply:
XGBTuner with `feature_type="itervar"` has the restriction that the AST tree should keep the same structure for all schedule in the search space. So we can get the fix-length feature vector for all schedule in the search space.

When I implemented the winograd schedule, I found different tile size will produce feature vectors with different lengths, so I didn't include it as a tunable knob.  

I expect the tuning will hang as @FrozenGene said, but it seems your change works well.. I will investigate later.

Reply:
@merrymercy ,
  That was my finding too for the ```xgb```, but it was triggered only by #2111 . Will look into depth at that time, and will open a discussion subject on this topic if will be necessary.
  I don't touch tutorials for this work.